### PR TITLE
Refine probe tool 'xcatmn'

### PR DESCRIPTION
**Pull request by hu-weihua**
What are this pull request do:

1. restructure the code of ``xcatmn``
2. Using new hierarchy template
3. fix bug #1725 ``xcatprobe xcatmn usage errors in help menu``

**NOTE** bug #1528 **is still not fixed. I will fix it soon**

Below output maybe helpful for review:

The output of usage:

    # xcatprobe xcatmn -h
    Usage:
    xcatmn -h
    xcatmn -i <install_nic> [-V]

    Description:
    After xCAT installation, use this command to check if xCAT has been installed correctly and is    ready for use.
    For hierarchical cluster, only the provision network on the same network as the management node is supported. If ...

    Options:
    -h : Get usage information of xcatmn
    -V : Output more information for debug
    -i : Required. Specify the network interface name of provision network on management node

The output of running check

    # xcatprobe -w new_xcatmn  -i eth0 -V
    <mn>: All xCAT deamons are running                                                                                [ OK ]
    <c910f02c04p04>: All xCAT deamons are running                                                                     [ OK ]
    <mn>: xcatd can receive command request                                                                           [ OK ]
    <c910f02c04p04>: xcatd can receive command request                                                                [ OK ]
    <mn>: 'site' table is configured correctly                                                                        [ OK ]
    <mn>: Provision network is configured correctly                                                                   [ OK ]
    <c910f02c04p04>: 'site' table is configured correctly                                                             [ OK ]
    <c910f02c04p04>: Provision network is configured correctly                                                        [ OK ]
    <c910f02c04p04>: 'passwd' table is configured correctly                                                           [FAIL]
    <mn>: 'passwd' table is configured correctly                                                                      [ OK ]
    <mn>: Important directory is configured correctly                                                                 [ OK ]
    <mn>: SELinux is disabled on current server                                                                       [ OK ]
    <mn>: HTTP service works well                                                                                     [ OK ]
    <mn>: TFTP service works well                                                                                     [ OK ]
    <mn>: DHCP service works well                                                                                     [ OK ]
    <mn>: Firewall is closed on current server                                                                        [ OK ]
    <mn>: The disk space is enough for xCAT to work                                                                   [WARN]
    <mn>: The free space of '/' is less than 12 G
    <mn>: The IP of master is a static IP address                                                                     [ OK ]
    <mn>: The dhcpd.leases file is less than 100M                                                                     [ OK ]
    <c910f02c04p04>: There isn't username or password for 'system' in 'passwd' table
    <c910f02c04p04>: Important directory is configured correctly                                                      [FAIL]
    <c910f02c04p04>: installdir '/install' isn't mounted from the management node
    <c910f02c04p04>: tftpdir '/tftpboot' isn't mounted from the management node
    <c910f02c04p04>: SELinux is disabled on current server                                                            [ OK ]
    <c910f02c04p04>: HTTP service works well                                                                          [FAIL]
    <c910f02c04p04>: HTTP service isn't ready on 10.2.4.4
    <c910f02c04p04>: TFTP service works well                                                                          [FAIL]
    <c910f02c04p04>: TFTP service isn't ready on 10.2.4.4
    <c910f02c04p04>: DHCP service works well                                                                          [FAIL]
    <c910f02c04p04>: There isn't 'dhcpd' deamon in current server
    <c910f02c04p04>: Firewall is closed on current server                                                             [ OK ]
    <c910f02c04p04>: The disk space is enough for xCAT to work                                                        [WARN]
    <c910f02c04p04>: The free space of '/' is less than 12 G
    ======================do summary=====================
    [MN]: Check on MN PASS.                                                                                           [ OK ]
        The disk space is enough for xCAT to work                                                                     [WARN]
            The free space of '/' is less than 12 G
    [SN:c910f02c04p04]: Check on SN c910f02c04p04 FAILED.                                                             [FAIL]
        'passwd' table is configured correctly                                                                        [FAIL]
            There isn't username or password for 'system' in 'passwd' table
        Important directory is configured correctly                                                                   [FAIL]
            installdir '/install' isn't mounted from the management node
            tftpdir '/tftpboot' isn't mounted from the management node
        HTTP service works well                                                                                       [FAIL]
            HTTP service isn't ready on 10.2.4.4
        TFTP service works well                                                                                       [FAIL]
            TFTP service isn't ready on 10.2.4.4
        DHCP service works well                                                                                       [FAIL]
            There isn't 'dhcpd' deamon in current server
        The disk space is enough for xCAT to work                                                                     [WARN]
            The free space of '/' is less than 12 G

### CODE REVIEW

File: xCAT-probe/subcmds/xcatmn [hunk 1]
```diff
@@ -52,201 +50,361 @@ Options:
...
+    my $rst = 0;
+    @$error_ref = ();
+    my @deamon_list = ("SSL listener", 
+                       "DB Access", 
+                       "UDP listener", 
+                       "install monitor",
+                       "Discovery worker",
+                       "Command log writer");
+
+    my $output = `ps aux|grep -v grep|grep xcatd`;
```
Comment (neo954) [inline xCAT-probe/subcmds/xcatmn]: I think use external command `ps` is not a good idea.

How about check the all the pid files, and check if all these processes exist.

    open FILE, "<", "/var/run/xcatd.pid";
    $pid = <FILE>;
    close FILE;
    $ret = kill -SIGZERO, $pid;
    # Check the return value here.

And then, do the same thing against `/var/run/xcat/cmdlogservice.pid`, `/var/run/xcat/mainservice.pid`,
`/var/run/xcat/installservice.pid`, `/var/run/xcat/udpservice.pid`.

Comment (hu-weihua) [inline xCAT-probe/subcmds/xcatmn]: We need to check 6 daemons but ``xcatd`` just maintains 4 ``.pid`` files. So even ``kill -0 pid`` is a good option, but we can't leverage pid files under ``/var/run/xcat``.

Yes, ``ps`` isn't the best solution, but it's a acceptable option.

File: xCAT-probe/subcmds/xcatmn [hunk 2]
```diff
@@ -52,201 +51,364 @@ Options:
...
+    my $rst = 0;
+    @$error_ref = ();
+    my @deamon_list = ("SSL listener",
+        "DB Access",
+        "UDP listener",
+        "install monitor",
+        "Discovery worker",
+        "Command log writer");
+
+    my $output = `ps aux|grep -v grep|grep xcatd`;
```
Comment (neo954) [inline xCAT-probe/subcmds/xcatmn]: use `ps` here is ugly.

File: xCAT-probe/subcmds/xcatmn [hunk 3]
```diff
@@ -255,210 +417,227 @@ sub do_main_job {
...
-        }
-        else {
-            probe_utils->send_msg($outputtarget, "f", "$msg");
+sub check_disk {
+    my $error_ref = shift;
+
+    my $rst = 0;
+    @$error_ref = ();
+
+    my %dir_expectedspace_list = ("/var" => 1, "/tmp" => 1, "/install" => 10);
```
Comment (neo954) [inline xCAT-probe/subcmds/xcatmn]: Should check `/tftpboot` as well.

**Comment by neo954** [comment]
I believe this patch is good enough to merge. Yeah.

**Comment by neo954** [comment]
I will merge this one.

**Comment by daniceexi** [comment]
@neo954 👍

**Pull request merged**